ECO-967: Feature/store order

sergeysikachev opened the pull request:
- Developer(s): @sergeysikachev

- Peer Reviewer:

- Ticket: URL_HERE

- Project PR: URL_HERE

- Academy PR: URL_HERE (for feature documentation and if needed migration guides and project integration guide)

#### Please confirm

- [ ] No new OS components - or they have been approved by legal department
- [ ] All new or heavily changed classes get an "A" rating (Scrutinizer), except Facades, Factories, QueryContainers and Tests

#### Documentation

- [ ] Functional documentation provided or in progress?
- [ ] Integration guide for projects provided or not needed?
- [ ] Migration guides for all contained majors provided or not needed?

#### Release Table

   Module                | Release Type         | Constraints         |
   :--------------------- | :------------------------ | :--------------------- |
   ArvatoRss              |       major                 |                       |

#### Release Notes

- Store Order functionality added.
- Risk Check functionality partially refactored during store order implementation.
-----------------------------------------

#### Bundle ArvatoRss

_**Major:** Backwards-compatible feature_

Def of done (by responsible developer):
- [ ] All changes are backward-compatible. Outdated code is marked as deprecated
- [ ] The change is isolated and not mixed with any cleanups or other changes
- [ ] New code fits to the architecture rules
- [ ] All new or changed business logic is covered by functional tests (in Zed business layer)

##### Change log

- Store Order functionality added.
- Risk Check functionality partially refactored during store order implementation.

Commit:
ECO-967. The first version of transfer declaration.

Commit:
ECO-967. Add storeOrder method to api adapter.

Commit:
ECO-967. Add converters for StoreOrder call and factory to born converters.

Commit:
ECO-967. Add return types to interface methods.

Commit:
ECO-967. Update license header.

Commit:
ECO-967. Update return type.

Commit:
ECO-967. Store order exception class.

Commit:
ECO-967. Fix method name.

Commit:
EC967. Fix return type.

Commit:
ECo-967. Extend store order response transfer with error data.

Commit:
ECO-967. Extend api adapter with store order call.

Commit:
ECO-967. Basic response converter for store order.

Commit:
ECO-967. Add mappers, facade method and factory methods.

Commit:
ECO-967. Add post check plugin for store order.

Commit:
ECO-967. Fix arguments.

Commit:
ECO-967. Add needed getters to module config.

Commit:
ECO-967. Add aspect mappers.

Commit:
ECO-967. Remove comment.

Commit:
ECO-967. Remove redundant use.

Commit:
ECO-967. Mapping refactoring.

Commit:
Update license badge.

Commit:
ECO-967. Remove todo comment.

Commit:
ECO-966. Adjust arvato rss related transfers structure inside quote.

Commit:
ECO-967. Fix phpcs and phpmd warnings.

Commit:
Update travis files.

Commit:
ECO-967. Update compoer autoload sections to psr-4.

Commit:
ECO-967. Add databuilder.xml.

Commit:
ECO-967. Add _support with several mocks.

Commit:
ECO-967. Add codeception.yml files.

Commit:
ECO-967. Remove legacy support folder.

Commit:
ECO-967. Remove legacy tests.

Commit:
ECO-967. Test changes.

 - Fixed converter classes.
 - Facade, mapper and converter tests refactored using databuilders.
 - Mapper tests have been created for storeOrder.

Commit:
ECO-967. Fix plugin namespace.

Commit:
ECO-967. Fix test namespaces and names.

Commit:
ECO-967. Replace Store::getInstance call with facade injection.

Commit:
ECO-967. Improve converters.

Commit:
ECO-967. Improve tests.

Commit:
ECO-967. Fix wsdl path for tests.

alex-galych commented on src/SprykerEco/Zed/ArvatoRss/Business/Api/Adapter/SoapApiAdapter.php:
```diff
@@ -73,17 +61,44 @@ public function performRiskCheck(ArvatoRssRiskCheckRequestTransfer $requestTrans
+            $responseTransfer = new ArvatoRssStoreOrderResponseTransfer();
+            $responseTransfer->setIsError(true);
+            $responseTransfer->setErrorMessage($exception->getMessage());
+            return $responseTransfer;
```
Empty line before return is missing.

alex-galych commented on src/SprykerEco/Zed/ArvatoRss/Business/Api/Converter/StoreOrderResponseConverterInterface.php:
```diff
@@ -0,0 +1,20 @@
+     *
+     * @return \Generated\Shared\Transfer\ArvatoRssStoreOrderResponseTransfer
+     */
+    public function convert(stdClass $response);
```
You can use \stdClass without use statement for your class

dereuromark commented on tests/SprykerEcoTest/Zed/ArvatoRss/_support/Mock/MoneyFacadeMock.php:
```diff
@@ -0,0 +1,36 @@
+
+class MoneyFacadeMock implements ArvatoRssToMoneyInterface
+{
+    const DECIMAL_VALUE = 100.00;
```
Constants usually one groups by the common part: VALUE_X vs VALUE_X etc.

dereuromark commented on tests/SprykerEcoTest/Zed/ArvatoRss/Business/Facade/RiskCheckFacadeTest.php:
```diff
@@ -0,0 +1,52 @@
+        $this->assertEquals(static::RESPONSE_STRING_FIElD_VALUE, $riskCheckResponse->getErrorMessage());
+        $this->assertEquals(static::RESPONSE_STRING_FIElD_VALUE, $riskCheckResponse->getResultCode());
+        $this->assertEquals(static::RESPONSE_STRING_FIElD_VALUE, $riskCheckResponse->getActionCode());
+        $this->assertEquals(static::IS_ERROR, $riskCheckResponse->getIsError());
```
That sounds like a code smell: Please dont const trivial things like false here, and instead use

    $this->assertFalse($...)

dereuromark commented on tests/SprykerEcoTest/Zed/ArvatoRss/Business/Facade/AbstractFacadeTest.php:
```diff
@@ -0,0 +1,33 @@
+    const IS_ERROR = false;
+    const RESPONSE_STRING_FIElD_VALUE = 'test';
+
+    protected $quote;
```
seems to miss a doc block about its type.

dereuromark commented on tests/SprykerEcoTest/Zed/ArvatoRss/Business/Api/Mapper/Aspect/OrderMapperTest.php:
```diff
@@ -0,0 +1,85 @@
+class OrderMapperTest extends AbstractMapperTest
+{
+    /**
+     * @const double DECIMAL_VALUE
```
float - and no repeating of the name DECIMAL_VALUE required

dereuromark commented on src/SprykerEco/Zed/ArvatoRss/Business/Api/Mapper/Aspect/BillingCustomerMapper.php:
```diff
@@ -0,0 +1,91 @@
+class BillingCustomerMapper implements BillingCustomerMapperInterface
+{
+    /**
+     * @const string DATE_FORMAT
```
Retype DATE_FORMAT is not needed here

dereuromark commented on src/SprykerEco/Zed/ArvatoRss/Business/Api/Mapper/Aspect/BillingCustomerMapper.php:
```diff
@@ -0,0 +1,91 @@
+    const DATE_FORMAT = 'Y-m-d';
+
+    /**
+     * @var \SprykerEco\Service\ArvatoRss\Iso3166ConverterServiceInterface $iso3166
```
$iso3166 is wrong and should be removed

sergeysikachev replied:
Could you please explain why?

dereuromark commented on src/SprykerEco/Zed/ArvatoRss/Business/Api/Converter/StoreOrderResponseConverter.php:
```diff
@@ -0,0 +1,24 @@
+     *
+     * @return \Generated\Shared\Transfer\ArvatoRssStoreOrderResponseTransfer
+     */
+    public function convert(stdClass $response)
```
Why stdClass and not a more concrete implementation or interface?

Commit:
ECO-967. Update tests.

Commit:
ECO-967. Add payment type constant file.

Commit:
ECO-967. Add payment type related constants.

Commit:
ECO-967. Add missed fields to store order request.

Commit:
ECO-967. Code style fixes.

Commit:
ECO-967. Fix const docblock.

dereuromark commented on src/SprykerEco/Zed/ArvatoRss/ArvatoRssConfig.php:
```diff
@@ -27,4 +27,16 @@ public function getAuthorization()
+    /**
+     * @param string $paymentMethod
+     *
+     * @return mixed
```
really mixed?

Commit:
ECO-967. Update const docblock

Commit:
ECO-967. Insert empty line before return.

Commit:
ECO-967. Fix return type.

Commit:
ECO-967. Remove variable name from docblock.

Commit:
ECO-967. Move store order specific setters from common mapper.

Commit:
ECO-967. Add spryker/sales dependency to communication layer.

Commit:
ECO-967. Remove post check plugin.

Commit:
ECO-967. Use order transfer for store order request. Update mappers.

Commit:
ECO-967. Change argument type.

Commit:
ECO-967. Rename constants file.

Commit:
ECO-967. Add success result code constant.

Commit:
ECO-967. Extend ArvatoRssStoreOrderResponse transfer.

Commit:
ECO-967. Extend store order response converter.

Commit:
ECO-967. Update constans file name in test.

Commit:
ECO-967. Fix bridge.

Commit:
ECO-967. Add transaction logger.

Commit:
ECO-967. Add oms condition and command for store order + transaction logging.

Commit:
ECO-967. Fix code style issues.

Commit:
ECO-967. Fix phpmd warnings.

Commit:
ECO-967. Fix transaction type for risk check logging.

Commit:
ECO-967. Fix tests.

Commit:
ECO-967. Remove redundant use statement.

Commit:
ECO-967. SoapApiAdapter refactoring.

- Separate call objects have been created for each api call.
- Persistance layer related things have been renamed.

Commit:
ECO-967. Rename query, used to get log data.

Commit:
ECO-967. Modify call constants.

Commit:
ECO-967. Remove store order response mapper.

Commit:
ECO-967. Refactoring of store facade order test.

Commit:
ECO-965. Fix code style.

Commit:
ECO-967. Fix travis config filename.

Commit:
ECO-967. Update config.dist.php.

Commit:
ECO-967. Map registerOrder parameter with customer's isGuest property.

Commit:
ECO-967. Fix missed use statement.

Commit:
ECO-967. Update test after refactoring registeredOrder mapping.

Commit:
ECO-967. Rename constants file to *Config.

Commit:
ECO-967. Fix wrong namespaces.

Commit:
ECO-967. Fix phpcs errors.

dereuromark commented on config/config.dist.php:
```diff
@@ -5,3 +5,6 @@
 $config[ArvatoRssConstants::ARVATORSS_CLIENTID] = '';
 $config[ArvatoRssConstants::ARVATORSS_AUTHORISATION] = '';
+$config[ArvatoRssConstants::ARVATORSS_PAYMENT_TYPE_MAPPING] = [
+    'invoice' => \SprykerEco\Shared\ArvatoRss\ArvatoRssApiConfig::INVOICE,
```
This magic string comes up a few times. Maybe it can be a class const in config too? For to be reused in tests as well.

sergeysikachev replied:
It's supposed to be a $payment->getPaymentMethod() return value. It depends on payment methods available int the system. I didn't find a constant responsible for this.
For example, I'm interested in values in $paymentMethods array here. *vendor/spryker/spryker/Bundles/DummyPayment/src/Spryker/Yves/DummyPayment/Handler/DummyPaymentHandler.php*
Probably some other approach should be used here, but for now I see it only the way I implemented.

Commit:
ECO-967. Fix namespace.

Commit:
ECO-967. Fix wrong variable name in test.

Commit:
ECO-967. Add transfer objects for Order and CustomerBilling mappers.

Commit:
ECO-967. Refactoring mappers.

Commit:
ECO-967. Fix register order mapping.

Commit:
ECO-967. Fix mapper tests.

Commit:
ECO-967. Add communication token field to order transfer.

ievgenvarava committed:
ECO-1017: Added shipping address to RiskCheck request, fixed tests.

ievgenvarava committed:
ECO-1017: Added testcase for DeliveryCustomerMapper.

ievgenvarava committed:
ECO-1017: Fixed code sniffer warnings

ievgenvarava committed:
ECO-1017: fixed code review suggestions.

ievgenvarava committed:
Merge pull request #7 from spryker-eco/feature/ECO-1017-shipping-address-in-risk-check

Feature/eco 1017 shipping address in risk check
